pkg/csi/service/vanilla/controller.go

```diff
@@ -411,36 +421,46 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 	topologyRequirement := req.GetAccessibilityRequirements()
 	if topologyRequirement != nil {
 		// Get shared accessible datastores for matching topology requirement.
-		if c.manager.CnsConfig.Labels.Zone == "" || c.manager.CnsConfig.Labels.Region == "" {
-			// If zone and region label (vSphere category names) not specified in
-			// the config secret, then return NotFound error.
-			return nil, logger.LogNewErrorCode(log, codes.NotFound,
-				"zone/region vsphere category names not specified in the vsphere config secret")
-		}
-		vcenter, err := c.manager.VcenterManager.GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
-		if err != nil {
-			return nil, logger.LogNewErrorCodef(log, codes.NotFound,
-				"failed to get vCenter. Err: %v", err)
-		}
-		tagManager, err := cnsvsphere.GetTagManager(ctx, vcenter)
-		if err != nil {
-			return nil, logger.LogNewErrorCodef(log, codes.NotFound,
-				"failed to get tagManager. Err: %v", err)
-		}
-		defer func() {
-			err := tagManager.Logout(ctx)
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ImprovedVolumeTopology) {
+			sharedDatastores, datastoreTopologyMap, err = c.topologyMgr.GetSharedDatastoresInTopology(ctx,
+				topologyRequirement)
+			if err != nil || len(sharedDatastores) == 0 {
+				return nil, logger.LogNewErrorCodef(log, codes.NotFound,
+					"failed to get shared datastores for topology requirement: %+v. Error: %+v",
+					topologyRequirement, err)
+			}
+		} else {
+			if c.manager.CnsConfig.Labels.Zone == "" || c.manager.CnsConfig.Labels.Region == "" {
+				// If zone and region label (vSphere category names) not specified in
+				// the config secret, then return NotFound error.
+				return nil, logger.LogNewErrorCode(log, codes.NotFound,
+					"zone/region vsphere category names not specified in the vsphere config secret")
+			}
+			vcenter, err := c.manager.VcenterManager.GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
 			if err != nil {
-				log.Errorf("failed to logout tagManager. err: %v", err)
+				return nil, logger.LogNewErrorCodef(log, codes.NotFound,
+					"failed to get vCenter. Err: %v", err)
+			}
+			tagManager, err := cnsvsphere.GetTagManager(ctx, vcenter)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.NotFound,
```

Comment:
Use `codes.Internal`

Reply:
This is the old code, I just moved it into the `else` block when FSS is not enabled. Anyways, I will change it.